ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs

```diff
@@ -241,18 +241,37 @@ private static string GetTrackDisplayName(StorageFile file, MidiFile midi)
 
         private void AddInstruments(MidiFile midi)
         {
-            foreach (var programChangeEvent in midi.GetTimedEvents().Select(e => e.Event).OfType<ProgramChangeEvent>())
+            var timedEvents = midi.GetTimedEvents();
+            var programChangedEvents = timedEvents.Select(e => e.Event).OfType<ProgramChangeEvent>().ToList<ChannelEvent>();
+            var events = programChangedEvents;
+
+            if (!programChangedEvents.Any())
+            {
+                events = timedEvents
+                    .Select(e => e.Event)
+                    .OfType<NoteOnEvent>()
+                    .ToList<ChannelEvent>();
+            }
+
+            foreach (var currentEvent in events)
             {
-                if (this.Instruments.ContainsKey(programChangeEvent.Channel))
+                var programNumber = new SevenBitNumber(1);
+
+                if (currentEvent is ProgramChangeEvent currentParsedEvent)
+                {
+                    programNumber = currentParsedEvent.ProgramNumber;
+                }
+
+                if (this.Instruments.ContainsKey(currentEvent.Channel))
                 {
-                    if (!this.Instruments[programChangeEvent.Channel].Contains(programChangeEvent.ProgramNumber))
+                    if (!this.Instruments[currentEvent.Channel].Contains(programNumber))
```

Comment:
If we don't want duplicates, a `HashSet` would likely have been better than a `List`. I realise you didn't change this now, but calling this out as I see it. Note, by the way, that the ordering in a `HashSet` can change so if you want the preserve the ordering, the `List` is the right data structure.

Reply:
I think maintaining the order of instruments is useful. We might want to extend this in the future to split different instruments per channel into different players too.

Reply:
Makes sense. Thanks for clarifying.